File: frontend/app/types/analyze.ts
export type SprayStatus = 'safe' | 'caution' | 'danger'

export type PrecipRisk = 'low' | 'medium' | 'high'

export interface AnalyzeLocation {
  lat: number
  lng: number
}

export interface WeatherSeriesPoint {
  date: string
  temp_c?: number | null
  precip_mm?: number | null
  wind_kmh?: number | null
  rh_percent?: number | null
}

export interface WeatherData {
  wind_speed_kmh: number
  relative_humidity_percent: number
  next_6h_precip_mm: number
  precip_risk: PrecipRisk
  series_7d: WeatherSeriesPoint[]
}

export interface SoilMoistureSeriesPoint {
  date: string
  value_percent: number
}

export interface SoilMoistureData {
  value_percent: number
  source: 'mock'
  series_7d: SoilMoistureSeriesPoint[]
}

export interface SoilTempSeriesPoint {
  date: string
  value_c: number
}

export interface SoilTemperatureData {
  value_c: number
  series_7d: SoilTempSeriesPoint[]
}

export interface SoilProperties {
  sand_percent?: number | null
  silt_percent?: number | null
  clay_percent?: number | null
  bulk_density_kg_m3?: number | null
  organic_carbon_g_kg?: number | null
}

export interface SprayInfo {
  status: SprayStatus
  reasons: string[]
}

export interface AgronomyData {
  ready_to_plant: boolean
  ready_to_plant_reason: string
  spray: SprayInfo
}

export type PhaseBucket = 'new' | 'waxing' | 'full' | 'waning'
export type ZodiacElement = 'Fire' | 'Earth' | 'Air' | 'Water'
export type CropCategory = 'Root' | 'Leaf' | 'Fruit' | 'Flower'

export interface MoonAlmanac {
  phase_name: string
  phase_bucket: PhaseBucket
  zodiac_sign: string
  zodiac_element: ZodiacElement
  lunar_cycle_progress: number
  illumination: number
}

export interface PhenologyAlmanac {
  mouse_ear_oak_likely: boolean
  dandelion_spring_likely: boolean
  gdd_base10_cumulative_ytd: number | null
}

export interface TraditionalAlmanac {
  moon: MoonAlmanac
  recommended_crop_category: CropCategory
  recommendation_badge_label: string
  phenology: PhenologyAlmanac
  modern_traditional_alignment: 'aligned' | 'caution'
  wise_council_tip: string | null
}

export interface AnalyzeResponse {
  location: AnalyzeLocation
  weather: WeatherData
  soil_moisture: SoilMoistureData
  soil_temperature: SoilTemperatureData
  soil_properties: SoilProperties
  agronomy: AgronomyData
  /** Present when the backend exposes the Traditional Almanac (OpenDSS analyze v2). */
  almanac?: TraditionalAlmanac
}

File: backend/app/schemas/analyze.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class AnalyzeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    lat: float = Field(..., ge=-90, le=90)
    lng: float = Field(..., ge=-180, le=180)


class Location(BaseModel):
    lat: float
    lng: float


class WeatherSeriesPoint(BaseModel):
    date: str
    temp_c: float | None = None
    precip_mm: float | None = None
    wind_kmh: float | None = None
    rh_percent: float | None = None


class WeatherData(BaseModel):
    wind_speed_kmh: float
    relative_humidity_percent: float
    next_6h_precip_mm: float
    precip_risk: Literal["low", "medium", "high"]
    series_7d: list[WeatherSeriesPoint]


class SoilMoistureSeriesPoint(BaseModel):
    date: str
    value_percent: float


class SoilMoistureData(BaseModel):
    value_percent: float
    source: Literal["mock"]
    series_7d: list[SoilMoistureSeriesPoint]


class SoilTempSeriesPoint(BaseModel):
    date: str
    value_c: float


class SoilTemperatureData(BaseModel):
    value_c: float
    series_7d: list[SoilTempSeriesPoint]


class SoilProperties(BaseModel):
    sand_percent: float | None = None
    silt_percent: float | None = None
    clay_percent: float | None = None
    bulk_density_kg_m3: float | None = None
    organic_carbon_g_kg: float | None = None


class SprayInfo(BaseModel):
    status: Literal["safe", "caution", "danger"]
    reasons: list[str]


class AgronomyData(BaseModel):
    ready_to_plant: bool
    ready_to_plant_reason: str
    spray: SprayInfo


PhaseBucket = Literal["new", "waxing", "full", "waning"]
ZodiacElement = Literal["Fire", "Earth", "Air", "Water"]
CropCategory = Literal["Root", "Leaf", "Fruit", "Flower"]


class MoonAlmanac(BaseModel):
    phase_name: str
    phase_bucket: PhaseBucket
    zodiac_sign: str
    zodiac_element: ZodiacElement
    lunar_cycle_progress: float = Field(..., ge=0.0, le=1.0)
    illumination: float = Field(..., ge=0.0, le=1.0)


class PhenologyAlmanac(BaseModel):
    mouse_ear_oak_likely: bool
    dandelion_spring_likely: bool
    gdd_base10_cumulative_ytd: float | None = None


class TraditionalAlmanac(BaseModel):
    moon: MoonAlmanac
    recommended_crop_category: CropCategory
    recommendation_badge_label: str
    phenology: PhenologyAlmanac
    modern_traditional_alignment: Literal["aligned", "caution"]
    wise_council_tip: str | None = None


class AnalyzeResponse(BaseModel):
    location: Location
    weather: WeatherData
    soil_moisture: SoilMoistureData
    soil_temperature: SoilTemperatureData
    soil_properties: SoilProperties
    agronomy: AgronomyData
    almanac: TraditionalAlmanac
